### PR TITLE
Jump to the next childDom if its `_dom` is null

### PR DESCRIPTION
**Pull request by Almo7aya**
I think this should fix https://github.com/developit/preact/issues/1343, but not other reconciliation issues like https://github.com/developit/preact/issues/1326

### CODE REVIEW

**Comment by andrewiggins** [comment]
Hi @Almo7aya! This is great! Thanks for taking the time to contribute a fix 🎉 

Could you add test case so we don't regress this in the future?

**Comment by marvinhagemeister** [comment]
Related: #1440

**Comment by Almo7aya** [comment]
@andrewiggins I'm stuck with writing a test case for this, the saveFocus function does not let me check for moving children, it saves the focused child and refocused it after rendering, so I can't detect if the children moved or not. Have any idea how to do this? :)

**Comment by marvinhagemeister** [comment]
@Almo7aya Let's use the code from #1343 as a test :tada: 

```jsx
it('should not re-render when a component returns undefined', () => {
  let Dialog = () => undefined;
  let updateState;
  class App extends Component {
    constructor(props) {
      super(props);
      this.state = { name: '' };
      updateState = () => this.setState({ name: ', friend' });
    }

    render(props, { name }) {
      return (
        <div>
          <Dialog />
          <h1 class="fade-down">Hi{name}</h1>
        </div>
      );
    }
  }

  render(<App />, scratch);
  clearLog();

  updateState();
  rerender();

  // We don't log text updates
  expect(getLog()).to.deep.equal([]);
});
```

**Comment by Almo7aya** [comment]
@marvinhagemeister Sorry!, but even this doesn't work, it always passes. The problem is children reconciler append children at the end of the parent not in their correct position this cause starting the animation, I cannot detect if the children were appended at the end or at their correct position using `insertBefore` 🤦‍♂️!!

**Comment by marvinhagemeister** [comment]
@Almo7aya are you sure? The test only passes with the changes from your PR. It fails in `master`.

**Comment by Almo7aya** [comment]
@marvinhagemeister I think I have some problems with my testing environment if you are sure that this fails in master and passes with this PR I'll commit your test case anyway?!

**Comment by marvinhagemeister** [comment]
@Almo7aya Good idea, we can always check if it runs properly in Travis here :+1:

**Comment by Almo7aya** [comment]
Just check it with travis it did fail with the previous changes
https://travis-ci.com/Almo7aya/preact/builds/105919357#L1317
https://github.com/Almo7aya/preact/blob/test-tesing/src/diff/children.js#L32

**Comment by marvinhagemeister** [comment]
@Almo7aya Oh right, I found why it always passes. Our test logger needs to be set up for it to work. We usually do it in the test runners `before` hook:

```js
import { logCall } from '../_util/logCall';

describe("render()", () => {
  // ...

  before(() => {
    logCall(Element.prototype, 'appendChild');
    logCall(Element.prototype, 'insertBefore');
    logCall(Element.prototype, 'removeChild');
    logCall(Element.prototype, 'remove');
  });

  // ...
});
```

Only once it's attached to a prototype it will start logging DOM operations.

**Comment by Almo7aya** [comment]
@marvinhagemeister That's great, thank you 🎉!

File: src/diff/children.js [hunk 1]
```diff
@@ -29,7 +29,15 @@ export function diffChildren(parentDom, newParentVNode, oldParentVNode, context,
 
 	let oldChildrenLength = oldChildren.length;
 
-	childDom = oldChildrenLength ? oldChildren[0] && oldChildren[0]._dom : null;
+	if (oldChildrenLength) {
```
Comment (marvinhagemeister) [inline src/diff/children.js]: Removing the length check around the `for` loop saves 2 bytes and all tests keep passing 🎉

File: src/diff/children.js [hunk 2]
```diff
@@ -29,7 +29,15 @@ export function diffChildren(parentDom, newParentVNode, oldParentVNode, context,
...
+		for (i = 0; i< oldChildrenLength; i++) {
```
Comment (marvinhagemeister) [inline src/diff/children.js]: nit: spacing: `i< oldChildrenLength` -> `i < oldChildrenLength`

**Comment by Almo7aya** [comment]
@marvinhagemeister  I think it's ready now. Thank you for your help, I didn't know about loggers functions in the test utils. Again thank you!🎉.

**Pull request merged**